tests/Authenticator.hs

```diff
@@ -0,0 +1,426 @@
+      -- NOTE: We always assume the user cooperates for now and choses the
+      -- first possible source
+      let (selectedKey, selectedCredential) = head credentialOptions
+
+      -- 8. Let processedExtensions be the result of authenticator extension
+      -- processing for each supported extension identifier → authenticator
+      -- extension input in extensions.
+      -- TODO: We don't suport extensions
+
+      -- 9. Increment the credential associated signature counter or the global
+      -- signature counter value, depending on which approach is implemented by
+      -- the authenticator, by some positive value. If the authenticator does
+      -- not implement a signature counter, let the signature counter value
+      -- remain constant at zero.
+      let updatedCounter = pkcsCounter selectedCredential + 1
+      let credentials = Map.update (\v -> Just $ v {pkcsCounter = updatedCounter}) selectedKey aCredentials
+
+      -- 10. Let authenticatorData be the byte array specified in § 6.1
+      -- Authenticator Data including processedExtensions, if any, as the
+      -- extensions and excluding attestedCredentialData.
+      -- NOTE: We don't just create the bytearray.
+      let rpIdHash = hash . encodeUtf8 $ M.unRpId rpId
+      let authenticatorData =
+            M.AuthenticatorData
+              { M.adRpIdHash = M.RpIdHash rpIdHash,
+                M.adFlags =
+                  M.AuthenticatorDataFlags
+                    { adfUserPresent = True,
+                      adfUserVerified = True
+                    },
+                M.adSignCount = updatedCounter,
+                M.adAttestedCredentialData = M.NoAttestedCredentialData,
+                M.adExtensions = Nothing,
+                M.adRawData =
+                  -- TODO: Use Put?
+                  BA.convert rpIdHash
+                    <> BS.singleton 0b00000101
+                    <> (toStrict . runPut $ Put.putWord32be . M.unSignatureCounter $ updatedCounter)
+              }
+
+      -- 11. Let signature be the assertion signature of the concatenation
+      -- authenticatorData || hash using the privateKey of selectedCredential
+      -- as shown in Figure , below. A simple, undelimited concatenation is
+      -- safe to use here because the authenticator data describes its own
+      -- length. The hash of the serialized client data (which potentially has
+      -- a variable length) is always the last element.
+      signature <-
+        PrivateKey.sign (pkcsPublicKey selectedCredential) (pkcsPrivateKey selectedCredential) (M.adRawData authenticatorData <> BA.convert (M.unClientDataHash clientDataHash))
+
+      -- 12. If any error occurred while generating the assertion signature,
+      -- return an error code equivalent to "UnknownError" and terminate the
+      -- operation.
+      -- NOTE: We don't produce any error
+
+      -- 13. Return
+      uh <- maybe (fail "UnknownError") pure $ pkcsUserHandle selectedCredential
```

Comment:
This needs to be looked into

Reply:
See the last note in https://www.w3.org/TR/webauthn-2/#authenticatorGetAssertion-return-values:
> Note: the returned [userHandle](https://www.w3.org/TR/webauthn-2/#public-key-credential-source-userhandle) value may be `null`, see: [userHandleResult](https://www.w3.org/TR/webauthn-2/#assertioncreationdata-userhandleresult).

So it should return a `Maybe UserHandle`